### PR TITLE
add update from the polonius wg

### PR DESCRIPTION
**Pull request by lqd**
This post is an Inside Rust update from the polonius working group describing our roadmap and milestones, and recent progress. This was co-written with @nikomatsakis.

Feedback is very much appreciated.

### CODE REVIEW

File: posts/inside-rust/2023-09-29-polonius-update.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+title: "Polonius update"
+author: Rémy Rakic and Niko Matsakis
+team: The Polonius Working Group <https://www.rust-lang.org/governance/teams/compiler#Polonius%20working%20group>
+---
+
+This post lays out a roadmap to try to get Polonius on stable by Rust 2024. It identifies some high-level milestones and summarizes the key goals, as well as the recent progress.
+
+## Background on polonius
+
+Polonius refers to a few things. It is a [new formulation](http://smallcultfollowing.com/babysteps/blog/2018/04/27/an-alias-based-formulation-of-the-borrow-checker/) of the borrow check. It is also a [specific project](https://github.com/rust-lang/polonius) that implemented that analysis, based on datalog. Our current plan does not make use of that datalog-based implementation, but uses what we learned implementing it to focus on reimplementing polonius within rustc.
```
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:12]:
```suggestion
Polonius refers to a few things. It is a [new formulation](http://smallcultfollowing.com/babysteps/blog/2018/04/27/an-alias-based-formulation-of-the-borrow-checker/) of the borrow checker. It is also a [specific project](https://github.com/rust-lang/polonius) that implemented that analysis, based on datalog. Our current plan does not make use of that datalog-based implementation, but uses what we learned implementing it to focus on reimplementing Polonius within rustc.
```

File: posts/inside-rust/2023-09-29-polonius-update.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            Some(v) => { temp = v; }
+            None => { }
+        }
+    }
+}
+```
+
+Here, the loan created on iteration N _may be used_ in the `None` path on iteration N+1, however the borrowed variable at that iteration is not the same.
+
+Issues like "NLL problem case #3", issue #47680 and others, were therefore deferred from NLLs, and left as future work, [polonius](http://smallcultfollowing.com/babysteps/blog/2018/04/27/an-alias-based-formulation-of-the-borrow-checker/).
```
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:57]:
```suggestion
Issues like "NLL problem case #3", issue #47680 and others, were therefore deferred from NLLs, and left as future work, [Polonius](http://smallcultfollowing.com/babysteps/blog/2018/04/27/an-alias-based-formulation-of-the-borrow-checker/).
```

File: posts/inside-rust/2023-09-29-polonius-update.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+Today, the trait solver produces higher-ranked outlives constraints and the borrow checker solves them. In the future, we would like to make the [next trait solver](https://github.com/rust-lang/trait-system-refactor-initiative/) responsible for solving these higher-ranked constraints itself, so that it only produces constraints on regions in the root [universe](https://rustc-dev-guide.rust-lang.org/borrow_check/region_inference/placeholders_and_universes.html#what-is-a-universe) (which the borrow checker can easily process). This would allow us to solve implication predicates like `for<T> { if (T: 'a, 'a: 'b) { T: 'b } }` without having to effectively reproduce the same trait solving logic again.
+
+In the shorter term, we can refactor the borrow checker to separate out the higher-ranked processing from the ordinary processing. The goal would be to preprocess the outlives constraints in a kind of polonius ["leak check"](https://rustc-dev-guide.rust-lang.org/traits/hrtb.html#basic-matching-and-placeholder-leaks), where we can compute the higher-ranked errors. This can then be removed once the trait solver can solve these constraints.
```
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:78]:
```suggestion
In the shorter term, we can refactor the borrow checker to separate out the higher-ranked processing from the ordinary processing. The goal would be to preprocess the outlives constraints in a kind of Polonius ["leak check"](https://rustc-dev-guide.rust-lang.org/traits/hrtb.html#basic-matching-and-placeholder-leaks), where we can compute the higher-ranked errors. This can then be removed once the trait solver can solve these constraints.
```

File: posts/inside-rust/2023-09-29-polonius-update.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+Current status: ⏳ members of the types team are starting to work on this task in the next few days.
+
+
+### 2. Location-insensitive loans in scope
+
+Out of the two key differences between polonius and the existing borrow check (regions as "sets of loans", and computing subtyping relations at each point in the CFG), this step is aimed at resolving the *first* difference, but not the second, so we call it the "location *in*sensitive loans in scope" (because subtyping is being done once, not per location): the idea can be described as "NLLs with the polonius model".
```
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:85]:
```suggestion
Out of the two key differences between polonius and the existing borrow check (regions as "sets of loans", and computing subtyping relations at each point in the CFG), this step is aimed at resolving the *first* difference, but not the second, so we call it the "location *in*sensitive loans in scope" (because subtyping is being done once, not per location): the idea can be described as "NLLs with the Polonius model".
```

File: posts/inside-rust/2023-09-29-polonius-update.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+Current status: ✅ we have completed prototypes, and have [an open PR](https://github.com/rust-lang/rust/pull/113218) to land this under a `-Z` flag, which should happen in the near future.
+
+### 3. Verify full test suite passes with location-insensitive polonius
+
+That PR does pass the full 15000+ tests in our suite, but we haven't yet checked on the crates published on crates.io with a crater run.
+
+Compared to our internal test suite, the vast majority of published crates are expected to build without errors. In that regard, it should be unlikely that issues would be found there, but it will be done regardless.
+
+Current status: ⏳ in-progress, should be finalized right when the PR is just ready to land.
```
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:101]: The wording here is a bit weird

File: posts/inside-rust/2023-09-29-polonius-update.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+Then the (harder-to-do-efficiently) work to incorporate location-sensitivity can start. This step will implement the first version of the analysis.
+
+At this point it can still be inefficient, and use the feature flag, but this is when the borrow checker should accept more expressive code than the current NLLs.
+
+Current status: we're in the design phase here, to adapt our datalog prototype and algorithms to rustc, imagining alternative ways to compute and propagate the subset constraints along the CFG.
+
+### 6. Model borrow checking and polonius in a-mir-formality
+
+The types team is building a model of Rust's MIR and trait system called [`a-mir-formality`][]. Once it reaches a sufficiently complete status, the intent is that the model will always be extended to cover new language features prior to stabilization. We are therefore working to add polonius into the model. This will in fact be the second time doing such modeling, as we already added polonius to a previous incarnation of a-mir-formality. In fact, that modeling work is what gave us the insights that enabled the location-insensitive polonius formulation now landing on nightly.
```
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:127]:
```suggestion
The Types team is building a model of Rust's MIR and trait system called [`a-mir-formality`][]. Once it reaches a sufficiently complete status, the intent is that the model will always be extended to cover new language features prior to stabilization. We are therefore working to add Polonius into the model. This will in fact be the second time doing such modeling, as we already added Polonius to a previous incarnation of a-mir-formality. In fact, that modeling work is what gave us the insights that enabled the location-insensitive polonius formulation now landing on nightly.
```

File: posts/inside-rust/2023-09-29-polonius-update.md [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+### 7. Location-sensitive pass stable
+
+In this milestone, we expect a lot of work on optimizations, and productization. 
+
+If a similar experience to NLLs in edition 2018 is to be expected again, another substantial amount of work and polish will also be needed to handle diagnostic differences and issues, ensuring errors and notes are clear enough, as well as the documentation.
+
+At this point, the location-sensitive pass is hopefully efficient enough, tested in practice, somewhat formally verified, and can be enabled in edition 2024.
+
+Around this time, librarification efforts can also be rebooted, to turn the in-tree polonius into a library, maybe using [Stable MIR][]. This is so that it could be reused elsewhere, for example in [rust-analyzer][], or [gccrs][], or by researchers working on verification tools (like [kani][]), [prusti][] and [creusot][]).
```
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:141]:
```suggestion
Around this time, librarification efforts can also be rebooted, to turn the in-tree Polonius into a library, maybe using [Stable MIR][]. This is so that it could be reused elsewhere, for example in [rust-analyzer][], or [gccrs][], or by researchers working on verification tools (like [kani][]), [prusti][] and [creusot][]).
```

Comment (jackh726) [reply posts/inside-rust/2023-09-29-polonius-update.md:141]: Also, missing links here.

File: posts/inside-rust/2023-09-29-polonius-update.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+[Stable MIR]: https://github.com/rust-lang/team/pull/729
+[rust-analyzer]: https://github.com/rust-lang/rust-analyzer
+[gccrs]: https://github.com/Rust-GCC/gccrs
+[kani]: https://github.com/model-checking/kani/
+[prusti]: https://github.com/viperproject/prusti-dev/
+[creusot]: https://github.com/xldenis/creusot
+
+## Conclusion
+
+We are very excited to see the plan for polonius coming into focus. At the moment, as we are still doing foundational work, we are not looking for volunteers or contributors unless they are well versed in the compiler. We do expect that as the project proceeds, there will be more and more need for new contributions. Stay tuned for updates!
```
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:152]:
```suggestion
We are very excited to see the plan for Polonius coming into focus. At the moment, as we are still doing foundational work, we are not looking for volunteers or contributors unless they are well versed in the compiler. We do expect that as the project proceeds, there will be more and more need for new contributions. Stay tuned for updates!
```

**Review by jackh726 — COMMENTED** [review]
Probably better to capitalize Polonius consistently, but not a huge deal I guess if its consistent.

**Comment by lqd** [comment]
We can consistently capitalize it here if we want, sure. I've pushed a commit to address this. (We often don't in practice, e.g. see Niko's [last post](https://smallcultfollowing.com/babysteps/blog/2023/09/22/polonius-part-1/)).

File: posts/inside-rust/2023-09-29-polonius-update.md (hunk 1 above)
Comment (lqd) [reply posts/inside-rust/2023-09-29-polonius-update.md]: I've pushed a commit to address this.

File: posts/inside-rust/2023-09-29-polonius-update.md (hunk 5 above)
Comment (lqd) [reply posts/inside-rust/2023-09-29-polonius-update.md]: I've pushed a commit to reword this into the following, which should be clearer:

> Current status: ⏳ in-progress, the crater run itself will be done just before the PR lands.

File: posts/inside-rust/2023-09-29-polonius-update.md (hunk 7 above)
Comment (lqd) [reply posts/inside-rust/2023-09-29-polonius-update.md]: I've pushed a commit for the capitalization. The links are just below the sentence though.

**Review by jackh726 — APPROVED** [review]

File: posts/inside-rust/2023-09-29-polonius-update.md (hunk 5 above)
Comment (jackh726) [reply posts/inside-rust/2023-09-29-polonius-update.md]: I probably would have removed the "just" here, but this is fine.

**Comment by jackh726** [comment]
@lqd did you want to wait until the 29th to land this? Given this is just inside rust, I'm happy to land whenever.

File: posts/inside-rust/2023-09-29-polonius-update.md (hunk 5 above)
Comment (lqd) [reply posts/inside-rust/2023-09-29-polonius-update.md]: I can do that.

**Comment by lqd** [comment]
@jackh726 Not particularly, I only picked "a week later" to allow time for review. Happy to change it to whichever date is preferred.

**Comment by jackh726** [comment]
Let's aim for tomorrow, Monday!

**Comment by lqd** [comment]
Sounds good. I pushed a commit to update the publication date.

File: posts/inside-rust/2023-09-25-polonius-update.md [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+### 7. Location-sensitive pass stable
+
+In this milestone, we expect a lot of work on optimizations, and productization. 
+
+If a similar experience to NLLs in edition 2018 is to be expected again, another substantial amount of work and polish will also be needed to handle diagnostic differences and issues, ensuring errors and notes are clear enough, as well as the documentation.
+
+At this point, the location-sensitive pass is hopefully efficient enough, tested in practice, somewhat formally verified, and can be enabled in edition 2024.
+
+Around this time, librarification efforts can also be rebooted, to turn the in-tree Polonius into a library, maybe using [Stable MIR][]. This is so that it could be reused elsewhere, for example in [rust-analyzer][], or [gccrs][], or by researchers working on verification tools (like [kani][]), [prusti][] and [creusot][]).
```
Comment (shepmaster) [inline posts/inside-rust/2023-09-25-polonius-update.md:141]:
```suggestion
Around this time, librarification efforts can also be rebooted, to turn the in-tree Polonius into a library, maybe using [Stable MIR][]. This is so that it could be reused elsewhere, for example in [rust-analyzer][], or [gccrs][], or by researchers working on verification tools (like [kani][], [prusti][] and [creusot][]).
```

Comment (lqd) [reply posts/inside-rust/2023-09-25-polonius-update.md]: Good catch, thank you.

File: posts/inside-rust/2023-09-25-polonius-update.md [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+## Milestones
+
+This is a rough roadmap, where we have the most visibility on the first steps:
+- each step has unknowns that will define what things need to be done in the later steps
+- therefore we're talking more about milestones for the longer term roadmap, and proper tasks for the shorter term.
+
+Here are the roadmap's milestones:
+
+![Graph of the Polonius roadmap](../../../../images/inside-rust/2023-09-25-polonius-update/roadmap.png)
```
Comment (calebcartwright) [inline posts/inside-rust/2023-09-25-polonius-update.md:72]: This isn't rendering for me on GitHub. Unsure if this is anticipated (e.g. just a quirk of paths that work on the site but not inline) but wanted to note in case it was typo

Comment (lqd) [reply posts/inside-rust/2023-09-25-polonius-update.md:72]: I don't believe images can render on github markdown previews because they are collected into a different web root `/static/` and this link pattern is used for resources in Inside Rust. 

Look at e.g. https://github.com/rust-lang/blog.rust-lang.org/pull/1103/files and https://github.com/rust-lang/blog.rust-lang.org/blob/92ac9e0679d4f41b338851aab1e18b33071bff9b/posts/inside-rust/2023-05-09-api-token-scopes.md?plain=1#L18 

FWIW the image does show up when locally serving the blog.

File: posts/inside-rust/2023-09-25-polonius-update.md [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+    loop {
+        match temp.maybe_next() {
+            Some(v) => { temp = v; }
+            None => { }
+        }
+    }
+}
+```
+
+Here, the loan created on iteration N _may be used_ in the `None` path on iteration N+1, however the borrowed variable at that iteration is not the same.
```
Comment (lcnr) [inline posts/inside-rust/2023-09-25-polonius-update.md:55]: how is the loan used in the `None` path? :think

File: posts/inside-rust/2023-09-25-polonius-update.md [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+Today, the trait solver produces higher-ranked outlives constraints and the borrow checker solves them. In the future, we would like to make the [next trait solver](https://github.com/rust-lang/trait-system-refactor-initiative/) responsible for solving these higher-ranked constraints itself, so that it only produces constraints on regions in the root [universe](https://rustc-dev-guide.rust-lang.org/borrow_check/region_inference/placeholders_and_universes.html#what-is-a-universe) (which the borrow checker can easily process). This would allow us to solve implication predicates like `for<T> { if (T: 'a, 'a: 'b) { T: 'b } }` without having to effectively reproduce the same trait solving logic again.
+
+In the shorter term, we can refactor the borrow checker to separate out the higher-ranked processing from the ordinary processing. The goal would be to preprocess the outlives constraints in a kind of Polonius ["leak check"](https://rustc-dev-guide.rust-lang.org/traits/hrtb.html#basic-matching-and-placeholder-leaks), where we can compute the higher-ranked errors. This can then be removed once the trait solver can solve these constraints.
```
Comment (lcnr) [inline posts/inside-rust/2023-09-25-polonius-update.md:78]: tbh I am a bit worried that this short term idea may not work out, idk if that means I would like to slightly reword this or just wait until we chat about it (hopefully today) :shrug:

Comment (nikomatsakis) [reply posts/inside-rust/2023-09-25-polonius-update.md:78]: Personally, I'd prefer not to wait to discuss -- I think that's putting too much weight into a status update blog post. But we could reword this to like "we are exploring refactoring the borrow checker" or something like that -- and "this could then be removed".

Comment (lcnr) [reply posts/inside-rust/2023-09-25-polonius-update.md:78]: yeah, changing it to "we're looking into refactoring" is also fine :+1: I mostly want to avoid definitive statements while not unsure whether this step is actually feasible

File: posts/inside-rust/2023-09-25-polonius-update.md (hunk 11 above)
Comment (lqd) [reply posts/inside-rust/2023-09-25-polonius-update.md:55]: yeah good spot, this actually describes a different minimization of the same issue 

```rust
struct Foo;

impl Foo {
    fn get_self(&mut self) -> Option<&mut Self> {
        Some(self)
    }

    fn bad_method(&mut self) {
        let mut var = self;
        
        loop {
            var = match var.get_self() {
                Some(v) => v,
                None => var
            };
        }
    }
}

fn main() {}
```

It's not the simplest/clearest of the 5 minimizations though, so I'll rewrite the description instead of changing the code example. I likely won't have time to do so until Wednesday/Thursday however.

File: posts/inside-rust/2023-09-25-polonius-update.md (hunk 12 above)
Comment (nikomatsakis) [reply posts/inside-rust/2023-09-25-polonius-update.md:78]:
```suggestion
In the shorter term, we are exploring refactoring the borrow checker to separate out the higher-ranked processing from the ordinary processing. The goal would be to preprocess the outlives constraints in a kind of Polonius ["leak check"](https://rustc-dev-guide.rust-lang.org/traits/hrtb.html#basic-matching-and-placeholder-leaks), where we can compute the higher-ranked errors. This could then be removed once the trait solver can solve these constraints.
```

File: posts/inside-rust/2023-09-25-polonius-update.md (hunk 11 above)
Comment (lqd) [reply posts/inside-rust/2023-09-25-polonius-update.md]: I tried to describe the cause succinctly, but it ends up less clear and less accurate as a simple soundbite. Niko rewrote it into the few sentences I've pushed in https://github.com/rust-lang/blog.rust-lang.org/pull/1147/commits/ce90cd698a56441486b8cd572ea8550454cd4560. 

It's clearer now, albeit wordy -- which could be fine on Inside Rust, or better suited to a footnote.

@lcnr what do you think?

File: posts/inside-rust/2023-09-25-polonius-update.md [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+- each step has unknowns that will define what things need to be done in the later steps
+- therefore we're talking more about milestones for the longer term roadmap, and proper tasks for the shorter term.
+
+Here are the roadmap's milestones:
+
+![Graph of the Polonius roadmap](../../../../images/inside-rust/2023-09-25-polonius-update/roadmap.png)
+
+### 1. Factoring out higher-ranked concerns from the main path
+
+Today, the trait solver produces higher-ranked outlives constraints and the borrow checker solves them. In the future, we would like to make the [next trait solver](https://github.com/rust-lang/trait-system-refactor-initiative/) responsible for solving these higher-ranked constraints itself, so that it only produces constraints on regions in the root [universe](https://rustc-dev-guide.rust-lang.org/borrow_check/region_inference/placeholders_and_universes.html#what-is-a-universe) (which the borrow checker can easily process). This would allow us to solve implication predicates like `for<T> { if (T: 'a, 'a: 'b) { T: 'b } }` without having to effectively reproduce the same trait solving logic again.
```
[the remark this answers is not in the document]
Comment (lqd) [reply posts/inside-rust/2023-09-25-polonius-update.md:80]: I can add that we can find additional details on terminology, on the problem itself, and on this future possible solution involving the trait solver, in this blog spot https://smallcultfollowing.com/babysteps/blog/2019/01/21/hereditary-harrop-region-constraints/.

**Review by lcnr — APPROVED** [review]

**Comment by lqd** [comment]
Now that lcnr's comment is resolved, I updated the publication date and squashed everything.
